### PR TITLE
Add IsTerminating property to ErrorRecord with tests

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
# PR Summary

An experiment at adding an `IsTerminating` property to `ErrorRecord`:

```
PS /Users/tyler/Code/PowerShell/PowerShell> throw "I'm terminating"

I'm terminating
At line:1 char:1
+ throw "I'm terminating"
+ ~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OperationStopped: (I'm terminating:String) [], RuntimeException
+ FullyQualifiedErrorId : I'm terminating

PS /Users/tyler/Code/PowerShell/PowerShell> $Error[0].IsTerminating
True

PS /Users/tyler/Code/PowerShell/PowerShell> Write-Error "I'm not terminating"

Write-Error "I'm not terminating" : I'm not terminating
+ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
+ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException

PS /Users/tyler/Code/PowerShell/PowerShell> $Error[0].IsTerminating
False
```

## PR Context

As a user, I want a clean and simple way to differentiate between terminating and non-terminating errors.

This PR helps bootstrap that effort by adding an `IsTerminating` property to `ErrorRecord` and setting where it makes sense.

This does _not_ change the formatting of an ErrorRecord. That is out of scope and requires a PM to approve what that will look like so in the meantime, this will provide the property that we can take advantage of later.

cc @vexx32 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

**Comment by iSazonov** [comment]
Much time has passed since that moment when I viewed the enhancement and as far as I remember there were a lot of unpredictable places where these terminating errors could appear from. I assume that this change covers too little of them for it to become a public property.

**Comment by vexx32** [comment]
In theory, every terminating error should still become an ErrorRecord before it reaches the user. If that is correct, it appears as though this should cover all of them.

There are some edge cases where the user is handed direct exceptions, but these are rare and should probably be fixed instead of being a blocker to something like this.

**Comment by iSazonov** [comment]
There is obvious case - ThrowTerminatingError() - see Add-Type cmdlet. 
If I remember correctly there is a lot of non-obvious cases.

**Comment by TylerLeonhardt** [comment]
@iSazonov, I fail to see why `ThrowTerminatingError` is an example. The default value on ErrorRecord is `IsTerminating` is `true` so that would be set as expected.

File: src/System.Management.Automation/engine/DefaultCommandRuntime.cs [hunk 1]
```diff
@@ -46,6 +47,15 @@ public DefaultCommandRuntime(List<object> outputList)
         /// <param name="errorRecord">Error record instance to process.</param>
         public void WriteError(ErrorRecord errorRecord)
         {
+            // Since this ErrorRecord was passed to WriteError,
+            // that means it's a non-terminating ErrorRecord.
+            // For RemotingErrorRecord's IsTerminating is already set
+            // so we shouldn't overwrite it.
```
Comment (vexx32) [inline src/System.Management.Automation/engine/DefaultCommandRuntime.cs]: I think this comment may be partially redundant, it's just describing what the code itself says. I don't think it's really needed. The bit about remoting is probably good to leave though.

File: src/System.Management.Automation/engine/MshCommandRuntime.cs [hunk 2]
```diff
@@ -2728,6 +2728,15 @@ private void DoWriteError(object obj)
             else
                 errorRecord.SetInvocationInfo(MyInvocation);
 
+            // Since this ErrorRecord was passed to WriteError,
+            // that means it's a non-terminating ErrorRecord.
+            // For RemotingErrorRecord's IsTerminating is already set
+            // so we shouldn't overwrite it.
```
Comment (vexx32) [inline src/System.Management.Automation/engine/MshCommandRuntime.cs]: Same thing here with the unneeded comment.

File: src/System.Management.Automation/engine/remoting/client/Job.cs [hunk 3]
```diff
@@ -769,6 +769,10 @@ private static Exception GetExceptionFromErrorRecord(ErrorRecord errorRecord)
 
         internal virtual void WriteError(ErrorRecord errorRecord)
         {
+            // Since this ErrorRecord was passed to WriteError,
+            // that means it's a non-terminating ErrorRecord.
```
Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/client/Job.cs]: This one could be removed completely.

File: src/System.Management.Automation/engine/remoting/client/Job.cs [hunk 4]
```diff
@@ -785,6 +789,10 @@ internal virtual void WriteError(ErrorRecord errorRecord)
 
         internal void WriteError(ErrorRecord errorRecord, out Exception exceptionThrownOnCmdletThread)
         {
+            // Since this ErrorRecord was passed to WriteError,
+            // that means it's a non-terminating ErrorRecord.
```
Comment (vexx32) [inline src/System.Management.Automation/engine/remoting/client/Job.cs]: Same here.

**Review by vexx32 — COMMENTED** [review]
Maybe a few unneeded comments but otherwise looks good overall.

We might want to do a more thorough review and ensure these are the only places that this needs to be set.

**Comment by iSazonov** [comment]
> I fail to see why ThrowTerminatingError is an example.

@TylerLeonhardt I mean follow although this is really not the case:

https://github.com/PowerShell/PowerShell/blob/a4e397697c3dbed414466398c4c695df712a5c7a/src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs#L1043-L1072

I reviewed code and my main concern is about patterns like `throw new someexception(errorRecord)`. Maybe there is something else but I can not remember.
You can find samples by "throw new .+error" pattern.
One is:
https://github.com/PowerShell/PowerShell/blob/a4e397697c3dbed414466398c4c695df712a5c7a/src/System.Management.Automation/engine/DefaultCommandRuntime.cs#L47-L53

**Comment by TylerLeonhardt** [comment]
@iSazonov I agree that all cmdlets should use `ThrowTerminatingError` instead of `throw` but I don’t think that really applies to my PR.

Those that just throw will be terminating errors and since `IsTerminating` defaults to `true` then there’s not really a problem here.

I might be misunderstanding you. Please let me know.

**Comment by iSazonov** [comment]
@TylerLeonhardt Thanks for clarify! It is encouraging. 

I think we need to update all our tests with "Should -Throw" to check IsTerminating property and to ensure that we don't skip something important.

I fast (not accurately) review (all) ErrorRecord uses (there is over 2200 references!) and now my optimism has gone down.
Below list where we could catch issues.

- StreamError Pipeline.Writer()
- private bool ProcessOutOfBand(PSObject so, bool isProcessingError) ???
- PSInvalidCastException -> ErrorRecord ??? As minimum it is best practice to explicitly set IsTerminating 
- commandRuntime._WriteErrorSkipAllowCheck ???
- internal RemoteException -> ErrorRecord ???
- CmdletProviderContext.WriteError - can throw, writeError and accumulate - 3 cases!
- Utils.TryRunAsImplicitBatch() - ps.AddCommand("Write-Error").AddParameter("InputObject", errorRecord).Invoke();  ??? - it is very interesting case in common - that Write-Error send - terminating or non-terminating? How resolve?
- DebuggerCommandProcessor.WriteErrorLine()  ???
- RunspaceOpenModuleLoadException - PSDataCollection<ErrorRecord> _errors;   ???
- PowerShell.cs - this.Streams.Error.Add(new ErrorRecord ... ???
- internal class ClientMethodExecutor - some methods like internal void Execute(Action<ErrorRecord> writeErrorAction)   ???
- Job.WriteError() and FailureErrorRecord property - not internally in the methods but how it is used in common   ???
- Job.ProcessJobFailure() - terminating or no?
- Job.cs and Job2.cs files ??? - need review how jobs work with terminating and non-terminating
- EncodeAndDecode.cs RemoteHostEncoder.cs ???
- ServerPowerShellDriver.cs private void HandleErrorDataAdded(object sender, DataAddedEventArgs e)  ???
- ServerRemotingProtocol2.cs internal void SendErrorRecordToClient(ErrorRecord errorRecord)  ???
- MiscOps.cs - need review all because:
  terminating?
```c#
                var errorRecord = new ErrorRecord(current.ErrorRecordToPass, current.ExceptionToPass);
                tuple.SetAutomaticVariable(AutomaticVariable.Underbar, errorRecord, context);
```
- HelpProvider.cs  internal void ReportHelpFileError(Exception exception, string target, string helpFile)  ??? Collect ErrorRecords - always terminating? HelpSystem has a lot of place where ErrorRecord is accumulated
- PSSnapInException - CreateErrorRecord() and ErrorRecord ???
- CommandNotFoundException - ErrorRecord ???
- CmdletInvocationException, ScriptCallDepthException, PipelineDepthException, ActionPreferenceStopException, PSArgumentException, PSArgumentNullException, PSArgumentOutOfRangeException, PSObjectDisposedException, ProviderInvocationException, SessionStateException, RuntimeException,  - ErrorRecord ???

File: src/System.Management.Automation/engine/DefaultCommandRuntime.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/DefaultCommandRuntime.cs]: Where set? In Serialization? What about previous versions and breaking remoting?

Comment (TylerLeonhardt) [reply src/System.Management.Automation/engine/DefaultCommandRuntime.cs]: > where set?

It's set in this PR. Look at `Job.cs`

I'll get back to you on the Remoting part.

**Comment by PaulHigin** [comment]
@TylerLeonhardt 
I disagree with the other PR comments that say the code comments should be reduced or removed, I feel they should remain.

However, I have concerns about whether this is a good and maintainable solution.  I feel it is too easy to miss places where ErrorRecords are written or collected.  For example do you cover the case where PowerShell collects error data when executing a pipeline?  Also whenever a new writer/collector is created, it needs to include the logic to set the 'IsTerminating' property.

I think it would be better to set the 'IsTerminating' property at the point where 'ErrorAction Stop' is processed, or where exceptions are processed so that the resulting ErrorRecord can be marked as terminating when it is created.  Can you look into this and see if it is possible to do?

**Comment by iSazonov** [comment]
There is over 2200 points where ErrorRecord is referenced but ~1077 for constructor. So we could add `IsTerminating` in the constructor parameters so that sets the value explicitly. Although there can be still a something specific in error streams/collectors.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee to discuss whether we can/should take an iterative approach to this

**Comment by TylerLeonhardt** [comment]
@PaulHigin good point. I probably need to use a `bool?` for this to handle remoting to downlevel versions. I’ll test this scenario and adjust accordingly.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we are ok with taking an iterative approach and fixing issues for completeness as they come up.  This requires that this be an experimental feature and not be out of experimental until we are confident the breadth of cases is covered.  We need to validate that addition to ErrorRecord type doesn't cause problem with remoting across PowerShell versions.

**Comment by TylerLeonhardt** [comment]
@PaulHigin and I had a chat offline on this PR and we agreed that this implementation is probably the right way to go. I investigated having the property be Non-Terminating by default and it doesn’t seem to be as straightforward as making it Terminating by default.

@SteveL-MSFT this change can’t be an experimental feature because it changes the type definition of the ErrorRecord class. `ExperimentalFeature` does not have the ability to expose/hide properties on types.

**Comment by TylerLeonhardt** [comment]
Also, I tested the remoting scenario for downlevel… My findings:

* With `Invoke-Command -Session $s`,  error handling is handled on the client side so remarkably, `IsTerminating` is set properly even if `Invoke-Command` is targeting a version of PowerShell without my change (my session targeted a pwsh 6.2 linux vm).
* Today, ErrorRecord is not deserialized when you do:
```pwsh
Invoke-Command -Session $s -ScriptBlock { Write-Error “foo” 2>&1 }
```
You get back a `PSCustomObject` so me adding `IsTerminating` should have no effect on deserializing.

**Comment by KirkMunro** [comment]
I meant to comment on this PR a while back but then got distracted, so my apologies if this is helpful and I'm surfacing it late in the discussion.

In my `HistoryPx` module, I identify any historical command as having encountered a terminating error simply by checking if the `ExecutionStatus` associated with the `HistoryInfo` record is equal to `System.Management.Automation.Runspaces.PipelineState.Failed`.

With that information, I find myself wondering if it is possible to set the `IsTerminating` flag on an `ErrorRecord` based on when that value would get set for a given command, rather than trying to cover the variety of scenarios where terminating errors can occur. I haven't looked into this possibility at all (I have my hands full with other tasks at the moment), but I wanted to toss the idea out there in case it helps.

**Comment by iSazonov** [comment]
@KirkMunro Perhaps we have the information in `public sealed class PipelineStateInfo`.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Error.Tests.ps1 [hunk 5]
```diff
@@ -66,6 +69,9 @@ Describe "Write-Error Tests" -Tags "CI" {
         #InvocationInfo verification
         $e.InvocationInfo | Should -Not -BeNullOrEmpty
         $e.InvocationInfo.MyCommand.Name | Should -BeNullOrEmpty
+
+        #IsTerminating verification
+        $e.IsTerminating | Should -BeFalse
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Write-Error.Tests.ps1]: Seems like there should be more explicit tests here.  Whether an Error is terminating or non-terminating depends on the state of that Error.  If it is handled, then it is considered non-terminating.  A non-terminating ErrorRecord can be terminating by changing ErrorAction, but then can be non-terminating again by using Trap with continue.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

**Comment by SteveL-MSFT** [comment]
@TylerLeonhardt can you summarize what the @PowerShell/powershell-committee needs to review?

**Comment by TylerLeonhardt** [comment]
@SteveL-MSFT  When a Non-Terminating error causes a script to stop (via `-ErrorAction Stop`) what should the behavior be.

**Comment by TylerLeonhardt** [comment]
Current implementation treats `ErrorAction` and `Termination` as two separate ideologies. I'm partial to the implementation.

**Comment by iSazonov** [comment]
@mklement0 Perhaps you have that say on @SteveL-MSFT request https://github.com/PowerShell/PowerShell/pull/10115#issuecomment-531844822

**Comment by TylerLeonhardt** [comment]
Still waiting on committee.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we believe the value would be to have both the initial terminating state as well as the resulting terminating state.  Suggestion is to add a member and every code path that can change the state (where `$?` could get set) should update that member.  ErrorRecord formatting can leverage this member to represent non-terminating errors as a different visual cue than terminating ones.

**Comment by TylerLeonhardt** [comment]
I'm just gonna close this. I don't have time to work on it anymore.

**Comment by iSazonov** [comment]
@vexx32 @mklement0 Have we a tracking issue for this? Could you please add a reference to the issue?

**Comment by vexx32** [comment]
Looks like it's #4781 I think?
